pkg/pbq/store/wal/segment.go

```diff
@@ -0,0 +1,324 @@
+//	+--------------------+------------------+-----------------+------------+
+//
+// We require the key-len because key is variadic.
+func encodeHeader(id *partition.ID) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
+	hp := headerPreamble{
+		S:    id.Start.UnixMilli(),
+		E:    id.End.UnixMilli(),
+		KLen: int16(len(id.Key)),
+	}
+
+	// write the fixed values
+	err := binary.Write(buf, binary.LittleEndian, hp)
+	if err != nil {
+		return nil, err
+	}
+
+	// write the variadic values
+	err = binary.Write(buf, binary.LittleEndian, []rune(id.Key))
+
+	return buf, err
+}
+
+func encodeEntry(message *isb.ReadMessage) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
+	body, err := encodeEntryBody(message)
+	if err != nil {
+		return nil, err
+	}
+	checksum := calculateChecksum(body.Bytes())
+
+	// Writes the message header
+	header, err := encodeEntryHeader(message, int64(body.Len()), checksum)
+	if err != nil {
+		return nil, err
+	}
+	wrote, err := buf.Write(header.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	if wrote != header.Len() {
+		return nil, fmt.Errorf("expected to write %d, but wrote only %d, %w", header.Len(), wrote, err)
+	}
+
+	// Writes the message body
+	wrote, err = buf.Write(body.Bytes())
+	if err != nil {
+		return nil, err
+	}
+	if wrote != body.Len() {
+		return nil, fmt.Errorf("expected to write %d, but wrote only %d, %w", header.Len(), wrote, err)
+	}
+	return buf, nil
+}
+
+func encodeEntryHeader(message *isb.ReadMessage, messageLen int64, checksum uint32) (*bytes.Buffer, error) {
+	watermark := message.Watermark.UnixMilli()
+
+	offsetInt, ok := message.ReadOffset.(isb.SimpleIntOffset)
+	if !ok {
+		return nil, fmt.Errorf("expect int64 offset for isb.ReadMessage")
+	}
+	offset, err := offsetInt.Sequence()
+	if err != nil {
+		return nil, err
+	}
+
+	entryHeader := entryHeaderPreamble{
+		WaterMark:  watermark,
+		Offset:     offset,
+		MessageLen: messageLen,
+		Checksum:   checksum,
+	}
+
+	// write the fixed values
+	buf := new(bytes.Buffer)
+	err = binary.Write(buf, binary.LittleEndian, entryHeader)
+	if err != nil {
+		return nil, err
+	}
+	return buf, nil
+}
+
+func encodeEntryBody(message *isb.ReadMessage) (*bytes.Buffer, error) {
+	m := new(bytes.Buffer)
+	enc := gob.NewEncoder(m)
+	err := enc.Encode(message.Message)
+	if err != nil {
+		return nil, fmt.Errorf("entry body encountered encode err: %w", err)
+	}
+	return m, nil
+}
+
+func calculateChecksum(data []byte) uint32 {
+	crc32q := crc32.MakeTable(IEEE)
+	return crc32.Checksum(data, crc32q)
+}
+
+// Write writes the message to the WAL. The format as follow is
+//
+//	+-------------------+----------------+-----------------+--------------+----------------+
+//	| watermark (int64) | offset (int64) | msg-len (int64) | CRC (unit32) | message []byte |
+//	+-------------------+----------------+-----------------+--------------+----------------+
+//
+// CRC will be used for detecting entry corruptions.
+func (w *WAL) Write(message *isb.ReadMessage) error {
```

Comment:
I think this error handling might get tricky. let's assume that the caller retries the write again, and we might write inconsistent data to WAL because of the previous incomplete write. Perhaps we should not return any error as it might complicate error handling. Retry on error till we can succeed?

Reply:
Good point, so I updated it to only bump the `woffset` when `Write` is actually succeed for atomic write so we are safe for retry.